components/supervisor/pkg/gitpod/gitpod-config.go

```diff
@@ -0,0 +1,244 @@
+		// inotify cannot watch inexistent file, let's poll
+		return nil
+	}
+	return service.watch(context)
+}
+
+func (service *ConfigService) tryStart() context.Context {
+	service.mutex.Lock()
+	defer service.mutex.Unlock()
+
+	if service.stop != nil {
+		return nil
+	}
+
+	log.WithField("location", service.location).Info("Starting watching...")
+	context, stop := context.WithCancel(context.Background())
+	service.stop = stop
+	return context
+}
+
+func (service *ConfigService) watch(context context.Context) error {
+	watcher, startErr := fsnotify.NewWatcher()
+	defer func() {
+		if startErr != nil {
+			log.WithField("location", service.location).WithError(startErr).Error("Failed to start watching...")
+		} else {
+			log.WithField("location", service.location).Info("Started watching")
+		}
+	}()
+	if startErr != nil {
+		return startErr
+	}
+
+	startErr = watcher.Add(service.location)
+	if startErr != nil {
+		watcher.Close()
+		return startErr
+	}
+
+	go func() {
+		defer log.WithField("location", service.location).Info("Stopped watching")
+		defer watcher.Close()
+
+		polling := make(chan struct{}, 1)
+		service.scheduleUpdateConfig(context, polling)
+		for {
+			select {
+			case <-polling:
+				return
+			case <-context.Done():
+				return
+			case err := <-watcher.Errors:
+				service.dispatchError(err)
+			case <-watcher.Events:
+				service.scheduleUpdateConfig(context, polling)
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (service *ConfigService) scheduleUpdateConfig(context context.Context, polling chan<- struct{}) {
+	service.mutex.Lock()
+	defer service.mutex.Unlock()
+	if service.timer != nil {
+		service.timer.Stop()
+	}
+	service.timer = time.AfterFunc(100*time.Millisecond, func() {
+		err := service.updateConfig()
+		if service.tryPolling(context, err) {
+			polling <- struct{}{}
+		} else if err != nil {
+			service.dispatchError(err)
+		}
+	})
+}
+
+func (service *ConfigService) dispatchError(err error) {
+	service.mutex.Lock()
+	defer service.mutex.Unlock()
+	for listener := range service.listeners {
+		listener.errors <- err
+	}
+}
+
+func (service *ConfigService) tryPolling(context context.Context, err error) bool {
+	if !os.IsNotExist(err) {
+		return false
+	}
+	go func() {
+		timer := time.NewTicker(2 * time.Second)
+		defer timer.Stop()
+
+		for {
+			select {
+			case <-context.Done():
+				return
+			case <-timer.C:
+			}
+
+			if _, err := os.Stat(service.location); !os.IsNotExist(err) {
+				service.watch(context)
+				return
+			}
+		}
+	}()
+	return true
+}
+
+func (service *ConfigService) updateConfig() error {
+	service.mutex.Lock()
+	defer service.mutex.Unlock()
+
+	config, err := service.parse()
+	service.config = config
+	for listener := range service.listeners {
+		listener.configs <- service.config
+	}
+	return err
+}
+
+func (service *ConfigService) parse() (*GitpodConfig, error) {
+	data, err := ioutil.ReadFile(service.location)
+	if err != nil {
+		return nil, err
+	}
+	data, err = yaml.YAMLToJSON(data)
```

Comment:
Not sure relying on a third-party library to solve our own lack of proper definitions is the way to go.
I'd much rather we have proper struct tags for `yaml` on `GitpodConfig`.
We could modify the struct tags after we generate them from the json schema. We have a process like that implemented for ws-daemon-api.

We have no idea if this library maps fields as we expect them to.